File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>

// Protopype of function to calculate checksumm
int sum(long rem, int x);

int main(void)
{
    long card = 0;
    long value = 0;
    int rem = 0;
    int check_sum = 0;
    int mult = 0;
    // Represent digital position
    int x = 1;
    bool end = 0;
    string answer = "INVALID";
    do
    {
        card = get_long("Supply credit card number for synthetic check: ");
    }
    while (card <= 0);
    
    while (!end)
    {
        value = card / 10;
        rem = card % 10;
        // Detect last digit
        if (value == 0)
        {
            if (card == 4 && (x == 13 || x == 16))
            {
                answer = "VISA";
                check_sum += sum(rem, x);
                break;
            }
            else if (card == 3 || card == 5)
            {
                check_sum += sum(rem, x);
                break;
            }
            else
            {
                break;
            }
        }
        // Not last digit
        else
        {
            // More then 2 digits left
            if (card > 99)
            {
                check_sum += sum(rem, x);
                card = value;
                x++;
                continue;
            }
            if ((card == 34 || card == 37) && x == 14)
            {
                answer = "AMEX";
                check_sum += sum(rem, x);
                card = value;
                x++;
                continue;
            }
            if ((card >= 51 && card <= 55) && x == 15)
            {
                answer = "MASTERCARD";
                check_sum += sum(rem, x);
                card = value;
                x++;
                continue;
            }
            if ((value == 4 && (x == 12 || x == 15)))
            {
                check_sum += sum(rem, x);
                card = value;
                x++;
                continue;
            }
            else
            {
                break;
            }
        }
    }
    // Check to exit the cycle through the break
    if (strncmp(answer, "INVALID", 1))
    {
        // Validating
        if ((check_sum % 10) != 0)
        {
            answer = "INVALID";
        }
    }
	
    printf("%s\n", answer);
}

// Function calculates checksumm based on 'second-to-last digit'
int sum(long rem, int x)
{
    int check_sum = 0;
    if (x % 2)
    {
        check_sum = rem;
    }
    else
    {
        int mult = rem * 2;
        if (mult > 9)
        {
            check_sum = mult - 9;
        }
        else
        {
            check_sum = mult;
        }
    }
    return check_sum;
}

File: pset6/credit/credit.py
from cs50 import get_string

while True:
    card = get_string("Number: ")
    if card.isdigit():
        break
lenght = len(card)
x = int(card[0])
y = int(card[:2])
even_check = False
summ = 0
vendor = ["VISA", "AMEX", "MASTERCARD"]
selector = 10
# Primary check
if (x == 4 and (lenght == 13 or lenght == 16)):
    Pri_check = True
    selector = 0
elif ((y == 34 or y == 37) and lenght == 15):
    Pri_check = True
    selector = 1
elif ((y in range(51, 56)) and lenght == 16):
    Pri_check = True
    selector = 2
else:
    Pri_check = False
if Pri_check:
    for i in range((lenght - 1), -1, -1):
        x = int(card[i])
        if even_check:
            mult = x * 2
            if mult > 9:
                mult = mult - 9
            summ += mult
        else:
            summ += x
        even_check = not even_check

    if (summ % 10) == 0:
        print(vendor[selector])
    else:
        print("INVALID")
else:
    print("INVALID")